Handle properly the LinesMesh intersectionThreshold

barroij opened the pull request:
This PR is an attempt to handle properly the `intersectionThreshold` of `LinesMesh`. This "extra-width" cannot be added to the `BoundingInfo` using the boundingBias` as it is currently, because the threshold is then scaled according to the world matrix compute the world BoundingInfo. This can lead to missleading world `BoundingBox` (if the scale is 1000 and the threshold is 0.1, the world bbox would get an extra extend of 100 in each direction).

To fix this issue, I removed the code that put this intersectionThreshold as boundingBias into the BoundingInfo. Instead, I integrate this `intersectionThreshold` in the `Ray.intersectsBox` and `Ray.intersectsSphere` with an extra parameter.

barroij commented on src/Mesh/babylon.abstractMesh.ts:
```diff
@@ -1335,13 +1336,11 @@ module BABYLON {
-                var pickedPoint = worldOrigin.add(worldDirection);
+                const world = this.getWorldMatrix();
+                const worldOrigin = Vector3.TransformCoordinates(ray.origin, world);
+                const direction = ray.direction.scaleToRef(intersectInfo.distance, Tmp.Vector3[0]);
```
prevent new Vector3

barroij commented on src/Mesh/babylon.geometry.ts:
```diff
@@ -120,11 +117,6 @@ module BABYLON {
-                    this.boundingBias = new Vector2(0, (<LinesMesh>mesh).intersectionThreshold);
-                    this._updateExtend();
-                }
-
```
stop changing the bias according to the intersectionThreshold

deltakosh commented on src/Culling/babylon.boundingBox.ts:
```diff
@@ -1,4 +1,7 @@
-﻿module BABYLON {
+module BABYLON {
+
+    const _identityMatrix = Matrix.Identity();
```
is const comparable to static readonly?

barroij replied:
nope, const just ensures at "compile time" that a variable is not changed

deltakosh replied:
ok so I would encourage using static readonly to save memory

barroij replied:
yup, I have no preference.
In both cases the Matrix is created only once. Having a closure on an unreachable and locally declared variable has the advantage of not "polluting" the class prototype.

I did it this way because that's what we were doing in BoundingSphere

deltakosh commented on src/Culling/babylon.boundingBox.ts:
```diff
@@ -93,7 +96,7 @@
             max.subtractToRef(min, this.extendSize).scaleInPlace(0.5);
 
-            this._update(worldMatrix || this._worldMatrix || Matrix.Identity());
+            this._update(worldMatrix || _identityMatrix);
```
What if I call it from the outside with reConstruct(min, max)? It will use identity instead of the current wm, right? This seems like an unwanted side effect

deltakosh commented on src/Culling/babylon.boundingBox.ts:
```diff
@@ -69,7 +72,7 @@
 
         /**
-         * Recreates the entire bounding box from scratch
+         * Recreates the entire bounding box from scratch, producing same values as if the constructor was called.
```
looks like a breaking change to me

barroij replied:
is it really unwanted as the comment said that it recreates the entine BBox from scratch ?
Also, from within the engine, it seems that the usual way to compute the world woordinates boundingBox after construction or reconstruction is to call update, that's why computing them from reconstruct is kind of extra-useless computation.

deltakosh replied:
It is a public API so we do not know if people use it differently hence we cannot change the way it works

barroij replied:
ok, I'll add another function that do what I want so that I don't change the API.

barroij replied:
Also it is strange that's the behavior `reConstruct` is not the same between BoundingBox and BoundingSphere

deltakosh commented on src/Culling/babylon.boundingBox.ts:
```diff
@@ -131,7 +135,7 @@
          */
         public setWorldMatrix(matrix: Matrix): BoundingBox {
-            this._worldMatrix.copyFrom(matrix);
+            this._worldMatrix = matrix;
```
Nope, you don't know if people will not change the matrix externally. copy has to be kept

barroij replied:
It is just strange to not change it in place when we call update, but change it in place when we call setWorldMatrix.
I mean, if one calls `setWorldMatrix` after an update, the matrix that will be modified will be one the one passed to the `update` function, which is the world Matrix of the Mesh. Do we really want to change the world Matrix of the change in this case ? (see our previous discussion here https://github.com/BabylonJS/Babylon.js/pull/5219#issuecomment-423580302)

barroij replied:
Actually this function should be removed IMHO. Having an accessor to it for debugging purpose is fine, but why a setter ? The only way to change it should to call `reConstruct` or `update`, as it would also update all the world coordinates stuff.

I guess some API change is allowed for a major version release such as 4.0, isn't it ? 😉

deltakosh commented on src/Culling/babylon.boundingBox.ts:
```diff
@@ -140,22 +144,29 @@
-                minWorld.minimizeInPlace(v);
-                maxWorld.maximizeInPlace(v);
+
+            if (world !== _identityMatrix) {
```
good idea

barroij replied:
yes, but can only work if this._worldMatrix is never modified in place :)

deltakosh commented on src/Culling/babylon.boundingBox.ts:
```diff
@@ -140,22 +144,29 @@
+                }
             }
 
-            // Extend
```
Why removing comments?

deltakosh commented on src/Culling/babylon.boundingInfo.ts:
```diff
@@ -167,9 +167,9 @@ module BABYLON {
-            let size = boundingBox.maximumWorld.subtract(boundingBox.minimumWorld);
-            return size.length();
+            const boundingBox = this.boundingBox;
+            const diag = boundingBox.maximumWorld.subtractToRef(boundingBox.minimumWorld, Tmp.Vector3[0]);
```
Good catch

deltakosh commented on src/Culling/babylon.ray.ts:
```diff
@@ -3,11 +3,13 @@ module BABYLON {
-        private _pvec: Vector3;
-        private _tvec: Vector3;
-        private _qvec: Vector3;
+        private static _edge1 = Vector3.Zero();
```
To save memory I would have preferred keeping them lazy created

deltakosh replied:
nevermind, I'm dumb

barroij commented:
@deltakosh Ok I reverted my changes to keep the API unchanged (even if some API change for a major release should be ok...)

Still, there is an inconherence between how the BoundingSphere and the BoundingBox behaves, although there are used exactly the same way and the API and the comments are the same. This is, at least, counter-intuitive. As I had did there, I would recommend to align BoundingBox on BoundingSphere, meaning that a call to `reConstruct` produces exactly same results as the constructor, and that there is no way to change the world matrix of the Bbox byt by calling `reConstruct` or `update`

I added a small change in the Matrix API so that the cached value `_isIdentity` can be properly set when the Identity matrix is constructed. I did not enable it by default because of the following 👍 The result of `isIdentity` might depend on a `considerAsTextureMatrix` parameter, which is pretty bad. It prevents a lot of optimisation when dealing with Identity matrix, as the identity-ness depends on external context which is not stored into the Matrix. So, if a Matrix has its `_isIdentity` to true, there is no way to figure whether it was computed with `considerAsTextureMatrix` to true or false. I would higly recomment to remove that and find another way to deal with 3x2 matrix. Moreover, the code seems to be just wrong : 
````javacsript
        public isIdentity(considerAsTextureMatrix = false): boolean {
            if (this._isIdentityDirty) {
                this._isIdentityDirty = false;
                if (this.m[0] !== 1.0 || this.m[5] !== 1.0 || this.m[15] !== 1.0) {
                    this._isIdentity = false;
                } else if (this.m[1] !== 0.0 || this.m[2] !== 0.0 || this.m[3] !== 0.0 ||
                    this.m[4] !== 0.0 || this.m[6] !== 0.0 || this.m[7] !== 0.0 ||
                    this.m[8] !== 0.0 || this.m[9] !== 0.0 || this.m[11] !== 0.0 ||
                    this.m[12] !== 0.0 || this.m[13] !== 0.0 || this.m[14] !== 0.0) {
                    this._isIdentity = false;
                } else {
                    this._isIdentity = true;
                }

                if (!considerAsTextureMatrix && this.m[10] !== 1.0) {
                    this._isIdentity = false;
                }
            }

            return this._isIdentity;
        }
````
the flag can be set to false even if value that are not in the 3x2 matrix are non-zero

deltakosh commented on src/Culling/babylon.boundingBox.ts:
```diff
@@ -93,7 +93,7 @@
             max.subtractToRef(min, this.extendSize).scaleInPlace(0.5);
 
-            this._update(worldMatrix || this._worldMatrix || Matrix.Identity());
+            this._update(worldMatrix || this._worldMatrix || Matrix.Identity(true));
```
That should be always true no?

barroij commented on src/Math/babylon.math.ts:
```diff
@@ -5075,35 +5082,48 @@ module BABYLON {
          * @param result defines the target matrix
          */
-        public static IdentityToRef(result: Matrix): void {
+        public static IdentityToRef(result: Matrix, markAsUpdated = false): void {
```
you're right, will remove it

deltakosh commented:
A tiny conflict and we can merge
